keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+successfully. This ensures that network-attached resource are available again
+for other Pods, including those that might get scheduled to other nodes. It
+also signals that it is safe to deallocate and delete the ResourceClaim.
+
+#### Communication between kubelet and resource kubelet plugin
+
+Resource kubelet plugins are discovered through the [kubelet plugin registration
+mechanism](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration). A
+new "ResourcePlugin" type will be used in the Type field of the
+[PluginInfo](https://pkg.go.dev/k8s.io/kubelet/pkg/apis/pluginregistration/v1#PluginInfo)
+response to distinguish the plugin from device and CSI plugins.
+
+Under the advertised Unix Domain socket the kubelet plugin provides the following
+gRPC interface. It was inspired by
+[CSI](https://github.com/container-storage-interface/spec/blob/master/spec.md),
+with “volume” replaced by “resource” and volume specific parts removed.
+
+##### `NodePrepareResource`
+
+This RPC is called by kubelet when a Pod that wants to use the
+specified resource is scheduled on a node.  The Plugin SHALL assume
+that this RPC will be executed on the node where the resource will be
+used.  The Plugin SHALL return device name and kind for allocated
+device[s].
+
+The Plugin SHALL ensure that there are json file[s] in CDI format
+for the allocated resource. These files SHALL be used by runtime to
+update runtime configuration before creating containers that use the
+resource.
+
+This operation SHALL do as little work as possible as it’s called
+after a pod is scheduled to a node. All potentially failing operations
+SHALL be done during allocation phase.
+
+This operation MUST be idempotent. If the resource corresponding to
+the `resource_id` has already been prepared, the Plugin MUST reply `0
+OK`.
+
+If this RPC failed, or kubelet does not know if it failed or not, it
+MAY choose to call `NodePrepareResource` again, or choose to call
+`NodeUnprepareResource`.
+
+On a successful call this RPC should return set of fully qualified
+CDI device names, which kubelet MUST pass to the runtime through the CRI
+protocol.
+
+```protobuf
+message NodePrepareResourceRequest {
+  // The UID of the ResourceClaim. This field is REQUIRED.
+  string resource_uid = 1;
+}
+
+message NodePrepareResourceResponse {
+  // These are the additional devices that kubelet must
+  // make available via the container runtime. A resource
+  // may have zero or more devices.
+  repeated string cdi_device = 1;
+}
+```
+
+CRI protocol MUST be extended for this purpose, e.g. list of CDI
+device ids should be added to the CRI Device structure:
+
+```protobuf
+// Device specifies a host device to mount into a container.
+message Device {
+    ...
+    string permissions = 3;
+    // Set of fully qualified CDI device names
+    repeated string cdi_device = 4;
+}
+```
+
+###### NodePrepareResource Errors
```

Comment:
do we fail the pod?

what does the user see when this happens?

Reply:
Same as with pods where publishing a volume fails: some event will have to be emitted for the pod with a description of the problem.